File: articles/mathematical-biology-logic-living-systems/python/stochastic_birth_death.py
"""
Stochastic birth-death workflows.

Run:
    python python/stochastic_birth_death.py
"""

from pathlib import Path

import pandas as pd

from math_biology_core import simulate_birth_death


ARTICLE_DIR = Path(__file__).resolve().parents[1]
SCENARIO_PATH = ARTICLE_DIR / "data" / "stochastic_scenarios.csv"


def main() -> None:
    scenarios = pd.read_csv(SCENARIO_PATH)

    rows = []

    for _, row in scenarios.iterrows():
        trajectory = simulate_birth_death(
            initial_population=int(row["initial_population"]),
            birth_rate=float(row["birth_rate"]),
            death_rate=float(row["death_rate"]),
            time_end=float(row["time_end"]),
            seed=int(row["seed"]),
        )

        final = trajectory[-1]

        rows.append(
            {
                "scenario": row["scenario"],
                "n_events": len(trajectory) - 1,
                "final_time": final["time"],
                "final_population": final["population"],
                "extinct": final["population"] == 0,
            }
        )

    print(pd.DataFrame(rows).round(4).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/mathematical-biology-logic-living-systems/r/stochastic_birth_death.R
# Stochastic birth-death workflow in R.

scenario_path <- file.path("data", "stochastic_scenarios.csv")
if (!file.exists(scenario_path)) {
  scenario_path <- file.path("..", "data", "stochastic_scenarios.csv")
}

simulate_birth_death <- function(initial_population, birth_rate, death_rate, time_end, seed) {
  set.seed(seed)

  time <- 0
  population <- initial_population
  rows <- data.frame(time = time, population = population, event = "initial")

  while (time < time_end && population > 0) {
    total_rate <- (birth_rate + death_rate) * population
    if (total_rate <= 0) break

    time <- time + rexp(1, rate = total_rate)

    if (time > time_end) break

    if (runif(1) < birth_rate / (birth_rate + death_rate)) {
      population <- population + 1
      event <- "birth"
    } else {
      population <- population - 1
      event <- "death"
    }

    rows <- rbind(rows, data.frame(time = time, population = population, event = event))
  }

  rows
}

scenarios <- read.csv(scenario_path)

summary_rows <- list()

for (i in seq_len(nrow(scenarios))) {
  s <- scenarios[i, ]

  sim <- simulate_birth_death(
    s$initial_population,
    s$birth_rate,
    s$death_rate,
    s$time_end,
    s$seed
  )

  final <- tail(sim, 1)

  summary_rows[[i]] <- data.frame(
    scenario = s$scenario,
    n_events = nrow(sim) - 1,
    final_time = final$time,
    final_population = final$population,
    extinct = final$population == 0
  )
}

summary_df <- do.call(rbind, summary_rows)
print(round(summary_df, 4))
